Figura Server side (or FSB)

lexize opened the pull request:
Main goal of this PR is make Figura even more extendable, make it so main backend is not dying from a flood of a new Figura users, and also make it easier for server to communicate with avatars.

What exactly this PR is going to add:
1. **Server side of Figura.** Useful for server owners, and safe enough for players. All the avatars' hashes will be checked both on server and then on client side, in order to be sure that server is sending exactly the avatar that has been requested. Host's avatars will be checked with ehashes (will explain what it is later), in order to check that host won't get unknown avatars equipped.
2. **EHashes and secret key.** E-Hash stands for "encrypted hash" (yes i know im very original and i mean this is a concept that will be used only inside project and players won't even need to care about it). It is used in order to check avatar ownership on host side. EHash is a hash-sum of avatar's hash, and hash of a secret key set in Figura's settings. In addition to basic hash checking client will check if incoming avatar ehash sent by a server equals to expected ehash. If it does, avatar will be allowed to be equipped, otherwise avatar won't be equipped and client will be notified about possible attempt of avatar replacement. P.S. In order to make uploaded avatar be allowed on multiple clients (for example, multiple instances or devices) client will have to set the same secret key in Figura's settings.
3. **Server events.** Like Bukkit events, used to extend server side of Figura yet still keep it safe enough. It can be used to provide a different source of avatars (for example, DB), allow or disallow client connection, and sending/receiving figura packets to/from client.
4. **Figura packets** A way for a Figura server and client sides to communicate. This can be used if your server implements something that could be listened and used within your Figura avatar, for example a global server event, or something like that. (This is advanced feature that uses FiguraBuffer as input or output).

Changes so far:
* Modified root gradle project and `gradle.properties` in order to apply mappings and include minecraft only in projects that are modded.
* Started writing server side of Figura, added bunch of packet classes.
* Added interface for platform independent FriendlyByteBuf, so all the needed functions can be also used in Bukkit plugin
* Added platform independent Identifier record, which is basically ResourceLocation, but both for Bukkit and Fabric/Forge.
* Started writing AvatarManager for server side.
* Added a little Events system to make figura server side extendable but also secure enough. There are several types of events: Regular ones, cancellable, and returnable. Returnable events can be used
for custom avatar/metadata/usedata loading implementation, for example from DB (by default everything including avatars, avatar metadata, and userdata, will be stored in a folder)

For now I'm gonna take a break from working on this project, so i am just making it a pull request to make it possible for maintainers to also make changes if they wish.

lexize committed:
The nyaboom

* Modified root gradle project and `gradle.properties` in order to apply mappings and include minecraft only in projects that are modded.
* Started writing server side of Figura, added bunch of packet classes.
* Added interface for platform independent FriendlyByteBuf, so all the needed functions can be also used in Bukkit plugin
* Added platform independent Identifier record, which is basically ResourceLocation, but both for Bukkit and Fabric/Forge.
* Started writing AvatarManager for server side, gonna rewrite a lot tho.
* Added a little Events system to make figura server side extendable but also secure enough. There are several types of events: Regular ones, cancellable, and returnable. Returnable events can be used
for custom avatar/metadata/usedata loading implementation, for example from DB (by default everything including avatars, avatar metadata, and userdata, will be stored in a folder)

From now i am going to take a big break from working on this, don't want to tell why but if short then i've got through something not good and just need some time.
I will get back to working on this project as soon as i can.

lexize committed:
FiguraServerAvatarManager Rewrite :meowoozy:

Rewritten most of avatar manager to make it look better and use less memory.
Added `copyBytes` method to Utils, using this to copy hash arrays in order to prevent them from being changed externally from events and such (i forgor how java works so i've almost let that happen :skull:)

lexize committed:
Max outcoming chunk size

PoolloverNathan commented:
Would it be possible to use Minecraft's existing chat signing key to sign avatars? This could be used to avoid needing to share the key manually, while also preventing other players (or the server, if it has the key) replacing avatars.

lexize commented:
Key won't be shared at all. It will be only used to check ownership of host avatars, and will be accessible only from host client. Sharing it means allowing a server to just replace your avatar without your permission. Also I suggest to just go to thread on discord cuz we are basically discussing everything FSB related in there.

About using this chat signature thing - I don't even know how it works internally.

lexize committed:
Little project restructure + tiny additions

I have been really burned out lately so i couldnt really add a lot, sorry :catplush:

lexize committed:
Little updates

Have been in code mood so decided to use this to work on FSB a bit.

lexize committed:
Small changes to FiguraUserManager

Made it at first return offline user, and only auth them when needed in handshake handler

lexize committed:
More changes

I am really sorry but i dont remember what exactly i changed here, i am very tired.

TheBunnyMan123 commented:
> Would it be possible to use Minecraft's existing chat signing key to sign avatars? This could be used to avoid needing to share the key manually, while also preventing other players (or the server, if it has the key) replacing avatars.

what if someone (like me) doesn't sign chat messages using something like no chat reports

lexize committed:
Deferred packets, some changes to FiguraUserManager and userdata fetch packet

lexize committed:
Started writing avatar uploading handling, also added Hash class.

* Started writing avatar uploading handling, will finish it later.
* Added Hash class, that represents an unchangeable (kinda, reflection says hi) hash. It's actual data can be got with Hash.get(), that will ***copy*** contained hash and put it into 32-byte long byte array.
* Started making all classes use Hash instead of `byte[]`

lexize committed:
Almost finished avatar uploading handling

lexize committed:
Finished working on avatar uploading handling

lexize committed:
Garbage collection

lexize committed:
Minor fix, forgot to uncomment one line

lexize committed:
Preparations before integrating with base Figura

lexize committed:
Started writing mod FSB implementation + some minor fixes

lexize committed:
Logging, gradle and fabric.mod.json changes

lexize committed:
Merge branch '1.20' of github.com:FiguraMC/Figura into fsb

lexize committed:
No more chunk size limit

lexize committed:
Minor fixes, a little bit of restructure.

lexize committed:
Some tweaks to NetworkStuff (very WIP)

lexize committed:
Some minor changes

lexize committed:
Made it work, at least part of it

I am officially saying, that this is a huge W

lexize committed:
A lot of changes

It actually works

lexize committed:
Major bug fixes

lexize committed:
Some fixes. Also added separate icon for FSB

Commit:
You forgor a little something

lexize committed:
Merge pull request #1 from ElectricDaFearox/fsb

Fixed the missing translation

lexize committed:
Another bugfix

lexize committed:
Visal bug fix

lexize committed:
I have no idea how i've missed this :sob:

lexize committed:
Added disconnect reason when connected to FSB

lexize committed:
Make FSB work only if server is not singleplayer

lexize committed:
Some fixes, some tweaks

lexize committed:
Spigot gaming (it actually works (i am really surprised))

lexize committed:
Update workflow for GitHub

lexize committed:
I have no idea how and why this happened-

lexize committed:
Fuck craftbukkit

lexize committed:
Exclude Spigot from subprojects and prevent github from uploading spigot artifacts

lexize committed:
I am stupid as always (sorry)

lexize committed:
Major bug fix, ServerPacketsAPI

lexize committed:
Removed unused imports in play networking handler

lexize committed:
Added side prefix to all paths, also made full List of used packets

lexize committed:
Some changes to Packet and Handlers

* Added full list of packets to Packet.PACKETS
* Added one more method to Handlers

lexize committed:
Status fix, notify packet

lexize committed:
This was the most stupid mistake i've ever made

lexize committed:
No async :pleading_face:

* Made everything in FSB sync, because async stuff have been making a lot of stuff unstable
* Rewrote some stuff blocking messages from backend if FSB is connected. Now it should have less problems.

lexize committed:
Change FSB to use platform specific stuff

Well, it *seems* to be working, though, i am, as always, not sure, there won't be any bugs
Tested both Fabric server and client, and Forge server and client

lexize committed:
Huh it actually works

lexize committed:
My attempts to fix FSB and Figura backend incompats

lexize committed:
This was the most stupid issue ever

lexize committed:
If there would be some stupid mistakes awards i for sure would win it

lexize committed:
AAAND THE WINNER OF THE "MOST STUPID MISTAKE IN CODE" AWARD ISSS LEEEXIIIIZEEEE CONGRATULATIONS

:catplant:

lexize committed:
Debug stuff

Now DEBUG variable in Spigot is not hardcoded. You can turn the debug mode on with -DfiguraDebug=true

lexize committed:
Merge branch '1.20' of github.com:FiguraMC/Figura into fsb

lexize committed:
No more deferred packets (they are not used at all)

lexize committed:
Metadata format change

Users can now have only 1 avatar equipped per time (backend consistency).

Changed format of metadata. You can safely start your server, and old format will be converted to a new one, but methods for this will eventually be removed.

lexize committed:
Side map

lexize committed:
Packets class, Forge networking mini rewrite (idk how to normally call it) (surprisingly it worked both on server and client wow)

lexize committed:
How did i miss it

lexize committed:
Another fix

lexize committed:
Translation string, utility functions, FSB tooltip details

lexize committed:
Commands preparations (WIP)

lexize committed:
meowmeowmeow

lexize committed:
Commands gaming (wip)

lexize committed:
Commands gaming

lexize committed:
Loading avatars received from FSB on main thread

lexize committed:
Fetch delay (to prevent avatars loading too fast which is breaking a lot of stuff)

lexize committed:
/fsb avatar clear <uuid>

lexize committed:
Badge command

lexize committed:
Made FSB work on main thread. Need to be tested on practice. (i really hope it actually works)

lexize committed:
Yipeeeeee pride badges

lexize committed:
Revert "Fetch delay (to prevent avatars loading too fast which is breaking a lot of stuff)"

This reverts commit b620ac34789b6d6bf39b1534cf2a20f6f699bea2.

lexize committed:
FSB IN THE SAME TIME WITH MAIN BACKEND LETS GOOOOOO

lexize committed:
Multiple upload modes

lexize committed:
Rebranding :meowoozy:

lexize committed:
Little fixes

lexize committed:
My stoopid is infinite

lexize committed:
Fixes

lexize committed:
Merge branch '1.20' of github.com:FiguraMC/Figura into fsb

lexize committed:
I am once again stupid asf

lexize committed:
Removed second UUID check in AvatarManager due to it being completely pointless

lexize committed:
Merge branch '1.20' of github.com:FiguraMC/Figura into fsb

UnlikePaladin commented:
the time has come

ElectricDaFearox commented:
"it's time to go to heaven my child"

UnlikePaladin reviewed (approved):
chonker PR, thanks Kate!
